### DISCUSSION TITLE
Include with variables

### DISCUSSION THREAD

**Discussion by cavo789** [quarto-dev/quarto-cli#8271] [Q&A] [answered]
### Description

Hi all

Very, very specific (perhaps strange) question about the shortcode `{{< include filename >}}` ([https://quarto.org/docs/authoring/includes.html](https://quarto.org/docs/authoring/includes.html))

Is it possible to run something like `{{< include filename vars="{servername=my-uat-server;ip=127.0.0.1}">}}` i.e. include a file and pass variables (here as a JSON string)) and, on a second line `{{< include filename vars="{servername=my-prod-server;ip=127.0.0.2}">}}` 

Right now, I'm using a `_variables.yml` file like below

```yaml
servers:
  uat:
    name: my-uat-server
    ip: 127.0.0.1
  prod:
    name: my-prod-server
    ip: 127.0.0.2
```

In my markdown file, I use the syntax `{{< var servers.uat.name >}}` and it works. Imagine a user guide on how to configure a server. There are a lot of lines in my markdown file and a lot of variables like `{{< var servers.uat.xxxx>}}`.  What if I wish exactly the same include but this time with `{{< var servers.prod.xxxx >}}`

I think I should have a `uat.qmd` and then, a `prod.qmd` file. Is it correct (perhaps something I don't know yet). The text will be the same in both files; except variables so my thought is: do you think we can include a file and pass variables during the include? Or any similar way to be able to have only one `.qmd` file and use dynamic variables.

Hope I'm clear enough.

Thanks!

**Comment by cderv** [comment, marked as answer]
It feels like [Project profiles](https://quarto.org/docs/projects/profiles.html) should be a way to go with this (as uat or prod being profile). Though [variables file](https://quarto.org/docs/authoring/variables.html) does not handle profiles yet. 

Are the two uat and prod part of the same rendering for project ? Profile and meta could already be used for that I think. 

But maybe profile are not the answer here.

> do you think we can include a file and pass variables during the include?

This would be a feature request. I don't about the best way to do that. Could be using the shortcode syntax... 🤔 that is a nice idea. 

Related discussion topic IMO
* https://github.com/quarto-dev/quarto-cli/discussions/5654

I definitely understand the need to not repeat yourself and reuse the .qmd with different variable. 

Side note: As quarto supports computation cells, you could also fill a .qmd as a template with different variables using a templating tool (like whisker), and then `{{< include ... >}}` the resulting filled template after the computation

**Reply by cavo789** [reply]
You're crazy fast! Thanks!!!

I'll take time to read and understand provided links.

Merci Christophe

**Comment by cscheid** [comment]
> > do you think we can include a file and pass variables during the include?
> 
> This would be a feature request. I don't about the best way to do that. Could be using the shortcode syntax... 🤔 that is a nice idea.

I'm sorry to be the bearer of bad news, but this is a _very_ hard feature to add, because of the interplay between the different passes of Quarto rendering. I can't foresee us implementing something like that in the next 2-3 years; it would require rewriting most of Quarto's internals.

**Reply by cavo789** [reply]
No problem at all Carlos!!!  A few years ago, I've developed my own set of Linux bash scripts in order to generate my documentation (and now I wish to abandon them and use Quarto; really more powerful and used by a large community).  In my own scripts, years ago, my *shortcodes* like *include* was indeed with an optional JSON content and, for my include script, it was possible to define variables.

So I've still my old way-of-doing-things in mind and try to make the match with features of Quarto.

Right now, I'm playing with projects and profiles like introduced by @cderv here above.  I think it's a really good idea to generate one PDF (book in term of Quarto) for deploying in user-acceptance (my use case right now) and a second for production.

Below my current work in progress:

```markdown
## Configure the Apache server

**Start a SSH connection on `{{< var servers.uat.name >}}` for UAT (`{{< var servers.prod.name >}}` for PROD)** 
and make sure the Apache configuration file is correct. Type `sudo cat /etc/httpd/conf.d/websites.conf` and 
verify it's equal to the source below.

Here is the content of the `/etc/httpd/conf.d/websites.conf` file (situation end of January 2024):

::: {.content-visible when-profile="UAT"}
DUMMY CONTENT - UAT
:::
::: {.content-visible when-profile="PROD"}
DUMMY CONTENT - PROD
:::
```

Ideally, I would be able to write **Start a SSH connection on `{{< var server.name >}}`** and, it'll be the name of the UAT server or PROD server depending on the profile.  

I continue to search how to achieve this...

(I've tried with files *_variables-UAT.yml* and *_variables-PROD.yml* but no, it's not working)

**Reply by cavo789** [reply]
Using `_environment-UAT` and `_environment-PROD` is working...

In `_environment-UAT`

```.env
SERVER_NAME=uat_server
```

Then in markdown: `**{{< env SERVER_NAME >}}**`

I should probably stop to use `_variables.yml` for values that depends on a Quarto profile.

Once again, many thanks for the so-fast answer and pointing me to the right direction (💯 )

**Reply by mcanouil** [reply]
For now `_variables.yml` is not profile dependent, follow:
- https://github.com/quarto-dev/quarto-cli/issues/5954

**Reply by cavo789** [reply]
Thanks for the confirmation. I'll thus migrate my *profile's depending* variables from `_variables.yml` to environment files.

**Reply by mcanouil** [reply]
Bonus, you should be able to use the environment variables also in code cells, for example in R with  `sys.getenv()`
